frontend/src/components/pages/RegistrantExperience/Waiver/index.tsx

```diff
@@ -51,16 +58,63 @@ const boldButtonTextStyles = {
 const marginBetweenSections = { sm: 7, md: 8, lg: 10 };
 
 interface WaiverPageProps {
+  nextBtnRef: React.RefObject<HTMLButtonElement>;
   waiverInterface: WaiverInterface;
   waiverDispatch: React.Dispatch<WaiverReducerDispatch>;
   campName: string;
 }
 
 const WaiverPage = ({
+  nextBtnRef,
   waiverInterface,
   waiverDispatch,
   campName,
 }: WaiverPageProps): React.ReactElement => {
+  const [isNameInvalid, setIsNameInvalid] = useState<boolean>(false);
+  const [isDateInvalid, setIsDateInvalid] = useState<boolean>(false);
+  const [
+    isRequiredClauseInvalid,
+    setIsRequriedClauseInvalid,
+  ] = useState<boolean>(false);
+  const [isOptionalClausesInvalid, setIsOptionalClausesInvalid] = useState<
+    boolean[]
+  >([]);
+
+  useEffect(() => {
+    // update the size of the isOptionalClausesInvalid accordingly if clauses are added or deleted
+    setIsOptionalClausesInvalid(
+      new Array(waiverInterface.optionalClauses.length).fill(false),
+    );
+  }, [waiverInterface.optionalClauses.length]);
+
+  useEffect(() => {
+    let nextBtnRefValue: HTMLButtonElement; // Reference to the next step button
+    const updateFormErrorMsgs = () => {
+      if (!checkName(waiverInterface.name)) setIsNameInvalid(true);
+      if (!checkDate(waiverInterface.date)) setIsDateInvalid(true);
+      if (!checkRequiredClauses(waiverInterface.agreedRequiredClauses))
+        setIsRequriedClauseInvalid(true);
+      const newIsOptionalClausesInvalid: boolean[] = Array.from(
```

Comment:
Could we rewrite this check for optional clauses? It's a little confusing as to what we are trying to accomplish here. From what i understand, we create an array of booleans where each boolean refers to if the optional clause is NOT AGREED. Then we set invalidity based on this array. I am not sure what the underlying intention for this is

Reply:
The variable `isOptionalClausesInvalid` stored an array of boolean where each Boolean refers to if the optional clause **has either been agreed or disagreed to**. Since the optional clauses are by default unfilled. We use `isOptionalClausesInvalid` to raise an error when clause is left unfilled. I'm not fully sure how to make the check clearer. The idea is that we create a new array of booleans by checking whether each clause is unfilled.